meta-mbl-apps/recipes-core/swupdate-handlers/swupdate-handlers.bb

```diff
@@ -0,0 +1,43 @@
+# Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: MIT
+
+SUMMARY = "Compiles arm swupdate handlers and hands the object files to swupdate so they can be linked with swupdate core"
+LICENSE = "BSD-3-Clause"
+
+LIC_FILES_CHKSUM = "file://${WORKDIR}/git/LICENSE.BSD-3-Clause;md5=1a8858961a0fa364bc79169ca26815db"
+SRC_URI = "\
+    ${SRC_URI_MBL_CORE_REPO} \
+"
+SRCREV = "${SRCREV_MBL_CORE_REPO}"
+SRCNAME = "swupdate-handlers"
+
+LIC_FILES_CHKSUM = "file://${WORKDIR}/git/LICENSE.BSD-3-Clause;md5=1a8858961a0fa364bc79169ca26815db"
+S = "${WORKDIR}/git/firmware-management/${SRCNAME}"
+
+inherit cmake
+
+EXTRA_OECMAKE += "\
+-DBOOTFLAGS_DIR=${MBL_BOOTFLAGS_DIR} \
+-DUPDATE_PAYLOAD_DIR=${MBL_SCRATCH_DIR} \
+-DLOG_DIR=${localstatedir}/log \
+-DROOTFS_TYPE=${MBL_ROOT_FSTYPE} \
+-DFACTORY_CONFIG_PARTITION=${MBL_FACTORY_CONFIG_DIR} \
+"
+
+# We fetch the swupdate headers and shove them into our handler's include dir, as our handlers depend on some
+# of swupdate's headers.
+do_fetch_headers() {
```

Comment:
I don't think this is the right approach.

Why we can't add `DEPENDS = "swupdate"`? This way bitbake will add all headers and generated binaries in the `${WORKDIR}/recipe-sysroot`.

Reply:
We need inter-task dependencies because:
* We compile swupdate-handlers to object files which need access to swupdate's headers
* We then move the swupdate-handlers object files to the swupdate sourcetree before swupdate's `do_compile` task runs, so the handlers can link with swupdate's "handler" kbuild module
* If we use DEPENDS we can't shove our handlers back into the swupdate sourcetree before its `do_compile` task runs IIUC

Reply:
I wonder if a cleaner way to do this would be to have a new `swupdate-headers` recipe that just installs headers into somewhere like `${includedir}/swupdate`. Then the `swupdate-handlers` recipe could just `DEPEND` on `swupdate-headers` and use `do_install` or `do_deploy` to make the MBL handlers available to the swupdate recipe.

I haven't read all of the recipe changes yet, but bypassing `do_deploy` and `do_install` to share things between recipes is hard to get right.

Reply:
@rwalton-arm, 

Thanks for the explanation.

I think this is not the right way of doing this. In this case, we should have everything concentrated in the swupdate bbappend.

Reply:
There are benefits to building our handlers outside of swupdate's build system:
* We can use extra warning flags when we compile.
* We can set up static analysis of our handlers.
* We can split code into multiple files without battling with swupdate's build system.
* We can potentially write handlers in C++ rather than C.

Reply:
We can still accomplish most of these tasks by having separate tasks functions.

This interdependency of the tasks between different recipes and files copy is not easy to maintain and looks weird.

Reply:
I'm not sure if littering the swupdate recipe with extra tasks is much better/clearer than having a separate recipe with inter task dependencies. Also, I don't particularly want to add extra compile flags to swupdate's build system as I think that's probably a can of worms we don't need to open at this point.

I would prefer it if I didn't have to copy the files to the deploy directory, and was hoping someone would have a suggestion which got me the same behaviour as `DEPENDS` (headers automatically available in the sysroot) using inter task dependencies. One possibility is to add our handlers as patches to swupdate, and use swupdate's build system as is, but that loses us the benefits Jonathan mentioned above. Another thing, in the future I thought we wanted to use swupdate more as a shared library as an `RDEPENDS` of UpdateD, so we can reuse stuff out of the swupdate core & write the more complex handlers in C++?

I'm not yet convinced that any of the suggestions in this thread are better than what I've got at the moment. Maybe just adding patches is the best approach, but then the next comment is going to be "why aren't we upstreaming the patches?" and I don't think our handlers are generic enough to be upstreamed at this point. I'll think about it and decide what I'm going to do in the morning...

Reply:
If we decide to keep it as a separate recipe, the @jonathanhaigh-arm suggestion of a `swupdate-headers` recipe is a good idea. And the `swupdate-handlers` recipe generates the object files needed by the main `swupdate` recipe. The main `swupdate` recipe then needs to `DEPENDS` on `swupdate-handlers` and at the do_configure or do_compile task you can copy the object files from the `recipe-sysroot` to the `handlers` folder.

Reply:
Just adding some ideas on how to set things in the `swupdate-handlers` recipe if we decide to follow what I suggested above.

You can have the following:
```
DEPENDS += "swupdate-headers"
do_install() {
    install -d ${D}/${libdir}/
    install -m 666 ${B}/CMakeFiles/swupdate-handlers.dir/*.c.o ${libdir}/
}
FILES_${PN} = ""
FILES_${PN}-dev += "${libdir}/*.c.o"
```

Reply:
One thing that Rob and I discussed yesterday is: if we have a `swupdate-headers` recipe, how do we keep the `SRCREV` consistent between the `swupdate` and `swupdate-headers` recipes?

Another thing that I've been thinking about is: in order to reduce the fragility of the recipes, we'll want to assume as little as possible about what the `swupdate` recipe actually does. One thing that concerns me is that we assume that swupdate's `do_configure` doesn't modify the header files that we use. We also assume things about where object files must be inserted into the swupdate build.

This might be controversial, but what about something like:
1. We have a swupdate-headers recipe that:
    * `requires` the swupdate recipe, ensuring that we're using the same `SRCREV` as swupdate.
    * Use the swupdate recipe's `fetch`, `patch` and `configure` tasks to make sure we're using properly patched and configured headers.
    * Has an empty `build` task.
    * Installs the headers into a standard place using an overridden `install` task, allowing the swupdate-handlers recipe to just `DEPEND` on `swupdate-headers` and then access them automatically using `#include <swupdate/blah.h>`.

1. In the swupdate-handlers code in mbl-core, we:
    * Create a proper library rather than just an object file.
    * Use CMake to install the the library into a standard location.
    This allows the swupdate recipe to just `DEPEND` on the swupdate-handlers recipe and link the handlers in a standard way.

1. In the swupdate bbappend, all we do is:
    * `DEPEND` on `swupdate-handlers`.
    * Arrange for `-lswupdate-handlers` to be passed to the linker.
    * Patch `handlers.c` (or whatever it's called) to register thin wrappers for our handlers. I.e.
      ```C
      int rootfsv4_handler_wrapper(struct img_type *img, void *data)
      {
          return rootfsv4_handler(img, data);
      }
      __attribute__((constructor))
      void rootfsv4_handler_init(void)
      {
          register_handler("ROOTFSv4"
                           , rootfsv4_handler
                           , IMAGE_HANDLER
                           , NULL);
      }
      ```

Just adding our handlers as `C` files into the swupdate recipe is temptingly simple, but having extra warnings from the compiler due to us building the rootfs handler separately has already caught bugs, so I'd be reluctant to give that up. I'm not really convinced one way or the other yet.

Reply:
@jonathanhaigh-arm,

Your proposal makes sense to me.

Reply:
Jonathan and I discussed this proposal earlier and I'm also on board.